### DISCUSSION TITLE
Can we use only 1 smart contract script for every trading pair?

### DISCUSSION THREAD

**Discussion by thuannguyen2010** [fallen-icarus/cardano-swaps#6] [Q&A] [answered]
I see that we need to specify offered/asked assets in spending script to help it knows which assets can be swapped.
Then I wonder if we can use 1 same spending script for every trading pair and it will read offered/asked assets from datum?

**Comment by fallen-icarus** [comment]
@thuannguyen2010 Yes, it is possible to push that information into the datum and use only a single spending script. It would require different logic for the beacons since there would need to be a beacon stored with every UTxO to tell them apart. I am not sure of all the trade-offs of this approach but I intend to try something like this in the next version.

**Comment by thuannguyen2010** [comment]
If we use a single spending script, we only need deploy spending script once. Then when users create swap, they don't need to create spending script again, just UTxO which contains price and swap data (asked/offered assets) (ada savings)

**Reply by fallen-icarus** [reply]
I'm not sure I follow. The latest version doesn't require deploying a spending script at all. Alice creates one UTxO as the tag UTxO and this UTxO only contains the address beacon; no reference script is part of this UTxO (in the latest version). All her other UTxOs only contain the swap price and swappable assets.

**Reply by fallen-icarus** [reply]
The latest version now only uses a single spending script. Each swap UTxO is individually tagged with the appropriate beacons. Now the spending script only needs to be stored on-chain once and everyone can reference it. I will be updating the [cardano-reference-script](https://github.com/fallen-icarus/cardano-reference-scripts) protocol so users can trustlessly share the reference script.

**Comment by fallen-icarus** [comment, marked as answer]
This is now how cardano-swaps works so I will be closing this discussion.
